imap_processing/ultra/l1b/de.py

```diff
@@ -24,35 +41,113 @@ def calculate_de(de_dataset: xr.Dataset, name: str) -> xr.Dataset:
     """
     de_dict = {}
 
-    # Placeholder for calculations
-    epoch = de_dataset.coords["epoch"].values
-
+    de_dataset = de_dataset.where(
+        de_dataset["START_TYPE"] != GlobalConstants.INT_FILLVAL, drop=True
+    )
     de_dict["epoch"] = de_dataset["epoch"]
-    de_dict["x_front"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["y_front"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["x_back"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["y_back"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["x_coin"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["tof_start_stop"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["tof_stop_coin"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["tof_corrected"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["eventtype"] = np.zeros(len(epoch), dtype=np.uint64)
-    de_dict["vx_ultra"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vy_ultra"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vz_ultra"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["energy"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["species"] = np.zeros(len(epoch), dtype=np.uint64)
-    de_dict["event_efficiency"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vx_sc"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vy_sc"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vz_sc"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vx_dps_sc"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vy_dps_sc"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vz_dps_sc"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vx_dps_helio"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vy_dps_helio"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vz_dps_helio"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["eventtimes"] = np.zeros(len(epoch), dtype=np.float64)
+    xf = get_front_x_position(
+        de_dataset["START_TYPE"].data,
+        de_dataset["START_POS_TDC"].data,
+    )
+
+    # Pulse height
+    ph_indices = np.where(
+        (de_dataset["STOP_TYPE"] == 1) | (de_dataset["STOP_TYPE"] == 2)
+    )[0]
+    ph_tof, ph_t2, ph_xb, ph_yb = get_ph_tof_and_back_positions(
+        de_dataset, xf, "ultra45"
+    )
+    ph_d, ph_yf = get_front_y_position(de_dataset["START_TYPE"].data[ph_indices], ph_yb)
+    ph_energy = get_energy_pulse_height(
+        de_dataset["STOP_TYPE"].data[ph_indices], ph_xb, ph_yb
+    )
+    ph_r = get_path_length((xf[ph_indices], ph_yf), (ph_xb, ph_yb), ph_d)
+    ph_ctof, ph_bin = determine_species_pulse_height(ph_energy, ph_tof, ph_r)
+    ph_etof, ph_xc = get_coincidence_positions(de_dataset, ph_tof, "ultra45")
+
+    # SSD
+    ssd_indices = np.where(de_dataset["STOP_TYPE"] >= 8)[0]
```

Comment:
Suggest changing all "where" to nonzero and making sure you only want the first occurrence or if you want all of the indices. (indices to me means there could be multiples whereas you're only grabbing the first ssd_index currently)
```suggestion
    ssd_indices = np.nonzero(de_dataset["STOP_TYPE"] >= 8)
```